src/include/miopen/where/problem_description.hpp

```diff
@@ -0,0 +1,110 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#pragma once
+
+#include "miopen/tensor_view_utils.hpp"
+#include <cstdint>
+#include <miopen/problem_description_base.hpp>
+#include <miopen/tensor.hpp>
+#include <miopen/activ.hpp>
+
+#include <string>
+
+namespace miopen {
+
+struct NetworkConfig;
+
+namespace where {
+
+enum class Direction
+{
+    Forward,
+    Backward,
+};
+
+struct BackwardProblemDescription : ProblemDescriptionBase
+{
+    // Backward constructor
+    BackwardProblemDescription(const TensorDescriptor& outputGradDesc_,
+                               const TensorDescriptor& conditionDesc_,
+                               const TensorDescriptor& inputGradDesc_,
+                               const TensorDescriptor& otherGradDesc_)
+        : direction(Direction::Backward),
+          outputGradDesc(outputGradDesc_),
+          conditionDesc(conditionDesc_),
+          inputGradDesc(inputGradDesc_),
+          otherGradDesc(otherGradDesc_)
+    {
+        if(!(isBroadcastable(inputGradDesc, otherGradDesc) &&
+             isBroadcastable(otherGradDesc, conditionDesc)))
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "WHERE::ProblemDescription: Dimensions of input, other, condition must be "
+                         "broadcastable.");
+        }
+    }
+
+    Direction GetDirection() const { return direction; }
+    const TensorDescriptor& GetOutputGradDesc() const { return outputGradDesc; }
+    const TensorDescriptor& GetConditionDesc() const { return conditionDesc; }
+    const TensorDescriptor& GetInputGradDesc() const { return inputGradDesc; }
+    const TensorDescriptor& GetOtherGradDesc() const { return otherGradDesc; }
+
+    bool IsSameType() const
+    {
+        if(inputGradDesc.GetType() != otherGradDesc.GetType() ||
+           inputGradDesc.GetType() != outputGradDesc.GetType())
```

Comment:
Looking at the backend kernel, input_grad and other_grad are optional, meaning they may not always be present. However, you wrote the code assuming that inputGradDesc and otherGradDesc exist.

Reply:
I am confused about the input_grad and other_grad options. I intended to represent an empty tensor gradient with dimensions {0}, but this is not valid for tensor descriptors in MIOpen. As a result, I can't find a way for users to pass an empty tensor in the API.

Reply:
Please note this,
https://github.com/ROCm/MIOpen/blob/847c979b5817fa0de0c1a143e9d12a4623b7156f/src/include/miopen/adam/problem_description.hpp#L108
You can do it another way.

Reply:
I have made input grad and other grad tensor optional

Reply:
When I first worked with optional tensors, it was difficult because I had nothing to refer to.
Thank you for your hard work too.
I'll just do a few tests and if there are no problems, I'll finalize your PR.